### PR TITLE
Fixing handling of filesystem errors

### LINKED ISSUE #1239: Unable to link supporting utilities
### Rancher Desktop Version

main (post 0.7.1)

### Rancher Desktop K8s Version

N/A

### What operating system are you using?

macOS

### Operating System / Build Version

MacOS Big Sur 11.6.2

### What CPU architecture are you using?

x64

### Windows User Only

_No response_

### Actual Behavior

When a supporting link doesn't have something already in the target location I'm seeing an error like...

```
Can't link to /usr/local/bin/docker: Error: ENOENT: no such file or directory, readlink '/usr/local/bin/docker'
```

Screen Shot 2022-01-13 at 10 00 05 AM

### Steps to Reproduce

Navigate to the supporting utilities page when something doesn't have a link or file in the target location for a supporting utilities

### Result

```
Can't link to /usr/local/bin/docker: Error: ENOENT: no such file or directory, readlink '/usr/local/bin/docker'
```

### Expected Behavior

There should be no error and it should be possible to link the supporting utility.

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by mattfarina**
Closes #1239

Errors from the filesystem do not have the properties of UnixError
such as stdout. This adds a second type check for the properties
found on filesystem errors in both mac and linux.

### CODE REVIEW

**Review by mook-as — DISMISSED** [review]
Looks fine to me, but we have a few instances of casting exceptions to `NodeJS.ErrnoException` — do we need to fix those too?

**Comment by mattfarina** [comment]
These errors are, I believe, `ErrnoException` so there may be a better way to handle this. Let me take a look. Thanks for the pointer.

**Comment by rak-phillip** [comment]
I originally made this change is because Typescript errors implicitly have a type of `unknown` so we need to assert that a given property exists before attempting access. In doing this, I made the mistake of assuming that anytime we attempt to access `err.code` that these errors would adhere to a similar structure..

For example, `k3sHelper.ts` only ever needed to assert that `err.code` exists before trying to access the message. The same goes for `lima.ts`, `unixlikeIntegrations.ts` (`code` or `message`), `pathConflictManager.ts`, and so on.

In short, I'm wondering if it might make sense to assert that only the properties that we're interested in exist before accessing them? Doing so would certainly undo any undefined behavior I may have introduced and allow us time to figure the correct types for the errors that we want to handle.

**Comment by mook-as** [comment]
Since we're pretty much just looking at `ex.code` everywhere, would something like this be useful instead?
```ts
export function errorHasCode(val: any, code: string): boolean {
  return val?.code === code;
}
```

We can then do something like:
```diff
-      if (isUnixFsError(ex) && ex.code !== 'ENOENT') {
+      if (!errorHasCode(ex, 'ENOENT')) {
         throw ex;
       }
```

Note that this would introduce a subtle behaviour change — in some cases we return (or throw) _only if_ it's a `UnixError` (but doesn't have a code we expect); this would now make it return (or throw) regardless of the error type, if it isn't the expected code.  This would have to be examined case-by-case and dealt with, since a universal behaviour wouldn't be sensible.

**Review by rak-phillip — APPROVED** [review]
LGTM

File: src/typings/unix.interface.ts [hunk 1]
```diff
@@ -11,3 +11,18 @@ export const isUnixError = (val: any): val is UnixError => {
...
+
+export interface NodeError {
+  syscall: string;
+  path: string;
+  code: string;
+  message: string;
+  errno: number;
+}
+
+export const isNodeError = (val: any): val is NodeError => {
```
Comment (mook-as) [inline src/typings/unix.interface.ts:23]: This can just be
```suggestion
export const isNodeError = (val: any): val is NodeJS.ErrnoException => {
```

**Review by mook-as — DISMISSED** [review]

File: src/typings/unix.interface.ts (hunk 1 above)
Comment (mattfarina) [reply src/typings/unix.interface.ts:23]: I did that and ended up with lint messages in the code. `NodeError` has `code` while `NodeJS.ErrnoException` has `code?`. It could be a string or undefined. `code` is used in places where undefined won't work. So, this is really looking for a `ErrnoException` with populated fields.

**Review by mook-as — APPROVED** [review]
